bcp47/library/Data/BCP47/Internal/Parser.hs

```diff
@@ -0,0 +1,21 @@
+module Data.BCP47.Internal.Parser
+  ( complete
+  ) where
+
+import Control.Applicative ((<|>))
+import Control.Monad (void)
+import Data.Text (Text)
+import Data.Void (Void)
+import Text.Megaparsec (Parsec, eof, lookAhead)
+import Text.Megaparsec.Char (char)
+
+-- | Ensure a specifier extends to the next '-' or end of input
+--
+-- Used for specifiers that can match some prefix of another specifier.
+-- For example, a @'Script'@ or @'Region'@ can accidentally be parsed
+-- from the prefix of a @'Variant'@
+--
+-- The alternative would be to use @'notFollowedBy'@ with knowledge of
+-- the legal characters in the next valid specifier.
+complete :: Parsec Void Text a -> Parsec Void Text a
+complete parser = parser <* lookAhead (void (char '-') <|> eof)
```

Comment:
Example:

```plaintext
  fromText:1:35:
    |
  1 | es-cKH-TNo-dQI-jYa-ikz-FOf-wdsW-CVKN9KJJ-jMQ9mRKf-H-CWayOY-o-ryapQev-x-B
    |                                   ^
  unexpected 'K'
  expecting '-'
```

Here we have the following specifiers:

```plaintext
lang: es
lang exts: cKH, TNo, dQI, jYa, ikz, FOf,
script: wdsW
region: ø
variants: CVKN9KJJ, jMQ9mRKf
extensions: H-CWayOY, o-ryapQev
private use: x-B
```

Without this combinator, we mistakenly parse the `Region` `CV` out of the `Variant` `CVKN9KJJ` because the `Region` parser doesn't look far enough ahead to see it's still in the middle of a specifier.

Instead, we can make sure each specifier looks ahead to see that the next char in the input is either `-` (indicating another specifier) or EOF.

Reply:
Nice find! Can we add tests for this, please? Definitely seems like a tricky case we'll want to make sure doesn't regress.